framework/math/ParallelVector/ParallelVector.h

```diff
@@ -0,0 +1,164 @@
+                 MPI_Comm communicator);
+
+  /**Copy constructor.*/
+  ParallelVector(const ParallelVector& other);
+
+  /**Move constructor.*/
+  ParallelVector(ParallelVector&& other) noexcept;
+
+  virtual std::unique_ptr<ParallelVector> MakeCopy() const = 0;
+  virtual std::unique_ptr<ParallelVector> MakeNewVector() const = 0;
+
+  /** Returns the raw stl-vector associated with the local data of this
+   * vector.*/
+  virtual std::vector<double>& RawValues() = 0;
+
+  /// Return the size of the locally owned portion of the parallel vector.
+  uint64_t LocalSize() const { return local_size_; }
+
+  /// Return the global size of the parallel vector.
+  uint64_t GlobalSize() const { return global_size_; }
+
+  /**
+   * Read only accessor to the entry at the given local index of
+   * the local vector.
+   *
+   * \note This accessor allows access to all locally stored elements,
+   *       including any data beyond local_size_ that may exist in derived
+   *       classes.
+   */
+  virtual double operator[](int64_t local_id) const = 0;
+
+  /**
+   * Read/write accessor to the entry at the given local index of
+   * the local vector.
+   *
+   * \note This accessor allows access to all locally stored elements,
+   *       including any data beyond local_size_ that may exist in derived
+   *       classes.
+   */
+  virtual double& operator[](int64_t local_id) = 0;
+
+  /// Return a vector containing the locally owned data.
+  virtual std::vector<double> MakeLocalVector() = 0;
+
+  /**
+   * Set the entries of the locally owned portion of the parallel vector
+   * to the given value.
+   */
+  virtual void Set(double value) = 0;
+
+  /**
+   * Set the entries of the locally owned portion of the parallel vector
+   * to the given STL vector.
+   */
+  virtual void Set(const std::vector<double>& local_vector) = 0;
+
+  /**
+   * Define a set or add operation for the given global id-value pair
+   *
+   * This routine adds the global id-value pair to the set operation cache,
+   * which upon execution of Assemble, communicates the operations to the
+   * appropriate process.
+   */
+  virtual void SetValue(int64_t global_id, double value, VecOpType op_type) = 0;
+
+  /**
+   * Group multiple operations into a single call.
+   *
+   * This routine goes through the given global id-value pairs and calls
+   * SetValue for each.
+   */
+  virtual void SetValues(const std::vector<int64_t>& global_ids,
+                         const std::vector<double>& values,
+                         VecOpType op_type) = 0;
+
+  /**In place adding of vectors. The sizes must be compatible.*/
+  virtual void operator+=(const ParallelVector& y) = 0;
+
+  /**Adds a vector multiplied by scalar a. Optimized for a=1.0 and -1.0*/
+  virtual void PlusAY(double a, const ParallelVector& y) = 0;
+
+  /**Sets the local values of one vector equal to another. The sizes must be
+   * compatible.*/
+  virtual void CopyValues(const ParallelVector& y) = 0;
```

Comment:
This would be better as an overload of `Set` in my opinion. Why introduce a new function name to do the same type of operation in a more particular way?

Reply:
IDK `x.Set(y)` has a literal feeling to it, for me. Set it to the same vector? Set it to the same value? What if x is a non-ghosted vector and y is ghosted, what will x be afterwards?

Reply:
I'm not a fan of having a gazillion different function names to do extremely similar things. In my opinion, it makes it harder in the documentation not easier. If someone wants to set a vector to something, it is useful to have all of the different ways a vector can be set next to one another. Does `x.Set(1.0)` feel like the vector becomes a scalar value to you? Why not have `x.SetAllValuesToScalar(1.0)` and `x.SetToSTLVector(y)`, and `x.SetToOtherParallelVector(y)`?

Reply:
I would change the name of this to `CopyLocalValues`.

Reply:
Done